frame/pallet-account-alias-registry/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+// This file is part of Noir.
+
+// Copyright (C) 2023 Haderech Pte. Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Account Alias Registry Pallet
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub mod weights;
+
+use crate::weights::WeightInfo;
+use codec::{Decode, Encode, MaxEncodedLen};
+use np_runtime::AccountName;
+pub use pallet::*;
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{LookupError, StaticLookup},
+	DispatchError, MultiAddress,
+};
+use sp_std::prelude::*;
+
+type AccountIdLookupOf<T> = <<T as frame_system::Config>::Lookup as StaticLookup>::Source;
+
+/// A provider for tag number that discriminates the same name accounts.
+pub trait TagProvider<T: Config> {
+	fn tag(id: &T::AccountId, name: &str) -> Result<u16, ()>;
+}
+
+/// A generator for ethereum address.
+pub trait EthAddressGenerator<T: Config> {
+	fn generate(id: &T::AccountId) -> Result<[u8; 20], ()>;
+}
+
+#[cfg_attr(feature = "std", derive(Hash))]
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Encode, Decode, MaxEncodedLen, TypeInfo)]
+pub enum AccountAlias {
+	AccountName(AccountName),
+	EthereumAddress([u8; 20]),
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	/// The module's config trait.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+		/// The provider for tag number that discriminates the same name accounts.
+		type TagProvider: TagProvider<Self>;
+		/// The generator for ethereum address.
+		type EthAddressGenerator: EthAddressGenerator<Self>;
```

Comment:
```suggestion
		/// The provider for tag number that discriminates the same name accounts.
		type AccountNameTagGenerator: AccountNameTagGenerator<Self>;
		/// The generator for ethereum address.
		type AccountIdToEthAddress: AccountIdToEthAddress<Self>;
```

What about more explicit names?